src/devices/bus/nes/nes_ines.hxx

```diff
@@ -643,8 +643,29 @@ void nes_cart_slot_device::call_load_ines()
 		mapper |= (header[8] & 0x0f) << 8;
 		// read submappers (based on 20140116 specs)
 		submapper = (header[8] & 0xf0) >> 4;
-		prg_size += ((header[9] & 0x0f) << 8) * 0x4000;
-		vrom_size += ((header[9] & 0xf0) << 4) * 0x2000;
+
+		// NES 2.0's extended exponential sizes, needed for loading 64MB PRG/CHR chunks. These bizarrely go up to 7 * 2^63!
+		auto expsize = [] (u8 byte) { return (2*(byte & 0x03) + 1) * std::pow(2.0, byte >> 2); };
```

Comment:
Can't this be done with shifting instead of introducing floating-point arithmetic?

Reply:
Is there a cross-platform int128_t available? The only reason I did it this way was to easily catch possible overflow. As noted the result can end up being 2^66 - 2^63. I suppose I could change it to conditionally check the shift size in byte, but this seemed shortest and easiest.

Reply:
There's no reason to support exponents that large, even if the hardware does in principle support such massive sizes. I believe MAME's various file loaders still don't fully support sizes of more than 4 GB, and your "256MB as an arbitrary limit" doesn't even go that far.

Reply:
Ugh, I just realized this comment here and the commit message are wrong too. Because of using 15 as a flag in the high nibble, this patch is needed to load PRG/CHR chunks >= 32MB.

Reply:
Yes, of course there's no reason to support a NES file that large. I'm just trying to make the code not break on edge cases where some dummy edits a header wrong. How do you suggest I proceed for error handling then? I need to both 1) error check possible stupidly large sizes in the header and 2) set an arbitrary upper limit (prg_size and chr_size themselves are uint32_t).

Reply:
Hey AJR, I thought about this issue a bit more. Again, to be clear the issue is that I need to correctly calculate and validate the size given in the header, not that I expect to use the preposterous potential sizes.

I only see two options. 1) as is using floating point because it's the only portable type that's wide enough.

Or 2) resort to a logarithmic comparison. If m,n are our coefficient and exponent from the header and we let 2^a be our fixed maximum size, we see that _m * 2^n > 2^a_  iff  _log2(m) + n > a_. Since we're dealing with _m = 1,3,5,7_ and we want to work strictly with integers we can instead use _ceil(log2(m)) + n > a_ as our condition. Now in terms of code this works out to checking something like `std::min(2*(byte & 0x03), 3) + (byte >> 2) > 28`, where I've let _a=28_— my arbitrary limit of 256MB—and the std::min expression was the simplest expression for _ceil(log2(m))_ I could come up with given `byte & 0x03` determines m.

My feeling is that having a check like `std::min(2*(byte & 0x03), 3) + (byte >> 2) > 28` rather than what I've submitted, `temp > 0x1p28`, is rather opaque. And it only buys us avoiding floats. To which I ask: why avoid the float? In this case the computation is done once (no compounded float errors), and the value is well within the bounds of a normalized IEEE double (only two bits of mantissa are even used). Anyway, let me know what you think.